be/src/types/bitmap_value.cpp

```diff
@@ -923,91 +923,197 @@ void BitmapValue::_convert_to_smaller_type() {
 }
 
 int64_t BitmapValue::sub_bitmap_internal(const int64_t& offset, const int64_t& len, BitmapValue* ret_bitmap) {
-    if (offset > 0 && offset >= _bitmap->cardinality()) {
-        return 0;
-    }
-    if (offset < 0 && std::abs(offset) > _bitmap->cardinality()) {
+    switch (_type) {
+    case EMPTY:
         return 0;
+    case SINGLE: {
+        if (offset >= 1 || offset < -1 || len <= 0) {
+            return 0;
+        } else {
+            ret_bitmap->add(_sv);
+            return 1;
+        }
     }
-    int64_t abs_offset = offset;
-    if (offset < 0) {
-        abs_offset = _bitmap->cardinality() + offset;
-    }
+    case SET: {
+        if ((offset > 0 && offset >= _set->size()) || (offset < 0 && std::abs(offset) > _set->size())) {
+            return 0;
+        }
+        int64_t abs_offset = offset;
+        if (offset < 0) {
+            abs_offset = _set->size() + offset;
+        }
 
-    int64_t count = 0;
-    int64_t offset_count = 0;
-    auto it = _bitmap->begin();
-    for (; it != _bitmap->end() && offset_count < abs_offset; ++it) {
-        ++offset_count;
+        std::vector set_values(_set->begin(), _set->end());
+        std::sort(set_values.begin(), set_values.end());
+        int64_t count = 0;
+        int64_t offset_count = 0;
+        auto it = set_values.begin();
+        for (; it != set_values.end() && offset_count < abs_offset; ++it) {
+            ++offset_count;
+        }
+        for (; it != set_values.end() && count < len; ++it, ++count) {
+            ret_bitmap->add(*it);
+        }
+        return count;
     }
-    for (; it != _bitmap->end() && count < len; ++it, ++count) {
-        ret_bitmap->add(*it);
+    default:
+        DCHECK_EQ(_type, BITMAP);
+        if ((offset > 0 && offset >= _bitmap->cardinality()) ||
```

Comment:
_bitmap->cardinality is a heavy operator. we should consider about save it as a temp variable